README.md
# fiber-http

A minimalist Golang web application for testing Opsani optimizations with high
throughput, predictable performance, and metrics instrumentation.

Built with [Fiber](https://docs.gofiber.io/) and
[FastHTTP](https://github.com/valyala/fasthttp).

## Exposed endpoints

* `/` - Returns a 200 (Ok) `text/plain` response of "move along, nothing to see
  here".
* `/metrics` - Metrics in Prometheus format for scraping.
* `/cpu{?operations,duration}` - Consume CPU resources for a given number of
  operations (fixed work), a given duration (fixed time), or whichever comes
  first. The `operations` query parameter is parsed as an unsigned integer
  with zero interpretted as unlimited and the `duration` parameter is parsed as
  a Golang Duration string. Default: `operations=0&duration=100ms`
* `/memory{?size}` - Consume memory resources by allocating a byte array of the
  given size (in human readable byte string format). Default: `size=10MB`
* `/time{?duration}` - Consume time by sleeping for the given duration (in
  Golang Duration string format). Default: `duration=100ms`
* `/request{?url}` - Proxy an HTTP GET request to a URL and return the status
  code & message body retrieved..

The resource endpoints of `cpu`, `memory`, and `time` are useful for triggering
the artificial consumption of resources for testing autoscale behaviors, error
handling, response to latency, etc.

The `request` endpoint enables testing of service dependencies and can be
reentrantly chained. For example, when running an instance locally on port 8480
a request made to
`http://localhost:8480/request?url=http://localhost:8480/time?duration=45ms`
would simulate an upstream service with a 45ms latency.

## Instrumentation

The application includes support for instrumentation with metrics systems for
testing optization with different measurement backends.

A [Prometheus](https://prometheus.io/) metrics endpoint is exposed at `/metrics`
that can be scraped.

[New Relic](https://newrelic.com/) instrumentation can be activated by setting
the `NEW_RELIC_LICENSE_KEY` environment variable to a valid New Relic license
key. The middleware will activate and log a status message after initialization.

Set `NEW_RELIC_APP_NAME` to define the corresponding New Relic APM identifier
(default: `fiber-http`).

## Listening ports

By default, the server listens on HTTP port 8480 and HTTPS port 8843 (N + 8400).
The port can be changed via the `HTTP_PORT` and `HTTPS_PORT` environment
variables, respectively.

## Initial memory allocation

An initial resident memory allocation be created by setting the
`INIT_MEMORY_SIZE` environment variable (in human readable byte string format,
e.g., `256MB`).

## Docker images

Docker images are published to [Docker
Hub](https://hub.docker.com/r/opsani/fiber-http).

```console
$ docker pull opsani/fiber-http:latest
```

Tasks for working with the container image are in the [Makefile](Makefile).

## License

Distributed under the terms of the Apache 2.0 license.


CHANGELOG.md
# CHANGELOG

fiber-http is a test application designed to simulate various workloads.

fiber-http is distributed under the terms of the Apache 2.0 license.

This changelog catalogs all notable changes made to the project. The format
is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/). Releases are
versioned in accordance with [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.2.0] - 2020-10-21

### Added

- The `/cpu` endpoint now accepts an `operations` query parameter for executing 
  a fixed work operation.

## [0.1.0] - 2020-09-18

Initial public release.

The `fiber-http` application provides a simple API driven interface for
simulating workloads by combing CPU utilization, memory consumption, and
latency.


main_test.go
// Copyright 2020 Opsani
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"io/ioutil"
	"net/http"
	"net/http/httptest"
	"regexp"
	"testing"
)

func TestRoot(t *testing.T) {
	app := newApp()
	req := httptest.NewRequest("GET", "http://localhost:8080", nil)

	resp, err := app.Test(req)
	if err != nil {
		t.Fatal(err.Error())
	}
	if resp.StatusCode != 200 {
		t.Errorf("expected status code: 200, got: %v", resp.StatusCode)
	}

	assertResponseBodyEquals(t, resp, "move along, nothing to see here")
}

func TestCPU(t *testing.T) {
	app := newApp()
	req := httptest.NewRequest("GET", "http://localhost:8080/cpu", nil)

	resp, err := app.Test(req)
	if err != nil {
		t.Fatal(err.Error())
	}
	if resp.StatusCode != 200 {
		t.Errorf("expected status code: 200, got: %v", resp.StatusCode)
	}
	assertResponseBodyMatches(t, resp, `consumed CPU for \d+ operations in 100\.\d+ms\n`)
}

func TestCPUOperations(t *testing.T) {
	app := newApp()
	req := httptest.NewRequest("GET", "http://localhost:8080/cpu?operations=5000&duration=5m", nil)

	resp, err := app.Test(req)
	if err != nil {
		t.Fatal(err.Error())
	}
	if resp.StatusCode != 200 {
		t.Errorf("expected status code: 200, got: %v", resp.StatusCode)
	}
	assertResponseBodyMatches(t, resp, `consumed CPU for 5000 operations in .*\n`)
}

func TestCPUOperationsDurationExpiresFirst(t *testing.T) {
	app := newApp()
	req := httptest.NewRequest("GET", "http://localhost:8080/cpu?operations=50000000&duration=100ms", nil)

	resp, err := app.Test(req)
	if err != nil {
		t.Fatal(err.Error())
	}
	if resp.StatusCode != 200 {
		t.Errorf("expected status code: 200, got: %v", resp.StatusCode)
	}
	assertResponseBodyMatches(t, resp, `consumed CPU for \d+ operations in 100\.\d+ms\n`)
}

func TestMemory(t *testing.T) {
	app := newApp()
	req := httptest.NewRequest("GET", "http://localhost:8080/memory", nil)

	resp, err := app.Test(req)
	if err != nil {
		t.Fatal(err.Error())
	}
	if resp.StatusCode != 200 {
		t.Errorf("expected status code: 200, got: %v", resp.StatusCode)
	}
	assertResponseBodyEquals(t, resp, "allocated 10.00MB (10485760 bytes) of memory\n")
}

func TestTime(t *testing.T) {
	app := newApp()
	req := httptest.NewRequest("GET", "http://localhost:8080/time", nil)

	resp, err := app.Test(req)
	if err != nil {
		t.Fatal(err.Error())
	}
	if resp.StatusCode != 200 {
		t.Errorf("expected status code: 200, got: %v", resp.StatusCode)
	}
	assertResponseBodyEquals(t, resp, "slept for 100ms\n")
}

func TestRequest(t *testing.T) {
	app := newApp()
	req := httptest.NewRequest("GET", "http://localhost:8080/request?url=http://opsani.com/", nil)

	resp, err := app.Test(req)
	if err != nil {
		t.Fatal(err.Error())
	}
	if resp.StatusCode != 200 {
		t.Errorf("expected status code: 200, got: %v", resp.StatusCode)
	}
}

func assertResponseBodyEquals(t *testing.T, resp *http.Response, expected string) {
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		t.Fatal(err.Error())
	}
	actual := string(body)
	if actual != expected {
		t.Errorf("expected %q but got %q", expected, actual)
	}
}

func assertResponseBodyMatches(t *testing.T, resp *http.Response, expectedPattern string) {
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		t.Fatal(err.Error())
	}
	actual := string(body)
	matched, err := regexp.MatchString(expectedPattern, actual)
	if !matched {
		t.Errorf("expected %q to match %q but it did not: %v", actual, expectedPattern, err)
	}
}


main.go
// Copyright 2020 Opsani
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"crypto/tls"
	"fmt"
	"log"
	"math"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"sync"
	"time"

	"github.com/ansrivas/fiberprometheus/v2"
	"github.com/gofiber/fiber/v2"
	"github.com/gofiber/fiber/v2/middleware/logger"
	"github.com/gofiber/fiber/v2/middleware/requestid"
	"github.com/inhies/go-bytesize"
	"github.com/valyala/fasthttp"

	"github.com/newrelic/go-agent/v3/newrelic"
)

var once sync.Once
var app *fiber.App
var initMemory []byte

func newApp() *fiber.App {
	once.Do(func() {
		app = fiber.New()
		app.Use(logger.New())
		app.Use(requestid.New())

		prometheus := fiberprometheus.New("fiber-http")
		prometheus.RegisterAt(app, "/metrics")
		app.Use(prometheus.Middleware)

		// activate New Relic if NEW_RELIC_LICENSE_KEY is in the environment
		if newrelicLicenseKey := os.Getenv("NEW_RELIC_LICENSE_KEY"); newrelicLicenseKey != "" {
			newrelicAppName := os.Getenv("NEW_RELIC_APP_NAME")
			if newrelicAppName == "" {
				newrelicAppName = "fiber-http"
			}
			newrelicApp, err := newrelic.NewApplication(
				newrelic.ConfigAppName(newrelicAppName),
				newrelic.ConfigLicense(newrelicLicenseKey),
			)
			if err == nil {
				app.Use(NewRelicMiddleware(newrelicApp))
				log.Println("New Relic middleware initialized")
			} else {
				log.Printf("WARNING: failed to initialize New Relic: %s\n", err)
			}
		}

		app.Get("/", func(c *fiber.Ctx) error {
			return c.SendString("move along, nothing to see here")
		})

		app.Get("/cpu", func(c *fiber.Ctx) error {
			operations, err := strconv.ParseUint(c.Query("operations", "0"), 10, 64)
			if err != nil {
				return err
			}
			duration, err := time.ParseDuration(c.Query("duration", "100ms"))
			if err != nil {
				return err
			}

			i := uint64(0)
			x := 0.0001
			start := time.Now()
			for time.Since(start) < duration {
				if operations != 0 && i == operations {
					break
				}
				x += math.Sqrt(x)
				i++
			}

			runtime := time.Since(start)
			return c.SendString(fmt.Sprintf("consumed CPU for %v operations in %v\n", i, runtime.String()))
		})

		app.Get("/memory", func(c *fiber.Ctx) error {
			size, err := bytesize.Parse(c.Query("size", "10MB"))
			if err != nil {
				return err
			}

			data := append([]byte{}, make([]byte, size)...)
			return c.SendString(fmt.Sprintf("allocated %v (%d bytes) of memory\n", size.String(), len(data)))
		})

		app.Get("/time", func(c *fiber.Ctx) error {
			duration, err := time.ParseDuration(c.Query("duration", "100ms"))
			if err != nil {
				return err
			}

			time.Sleep(duration)
			return c.SendString(fmt.Sprintf("slept for %v\n", duration.String()))
		})

		app.Get("/request", func(c *fiber.Ctx) error {
			remoteURL := c.Query("url")
			if remoteURL == "" {
				c.Status(fiber.StatusBadRequest)
				return c.SendString("error: missing required query parameter \"url\"")
			}

			client := fasthttp.Client{}
			statusCode, body, err := client.Get(nil, remoteURL)
			c.Status(statusCode)
			if err != nil {
				return err
			}
			return c.Send(body)
		})

		app.Use(func(c *fiber.Ctx) error {
			return c.SendStatus(fiber.StatusNotFound)
		})
	})

	return app
}

func main() {
	// Allocate an initial heap if requested
	if sizeEnv := os.Getenv("INIT_MEMORY_SIZE"); sizeEnv != "" {
		size, err := bytesize.Parse(sizeEnv)
		if err != nil {
			log.Fatal(err)
		}
		initMemory = append(initMemory, make([]byte, size)...)
		log.Printf("NOTICE: allocated %v (%d bytes) of memory\n", size.String(), len(initMemory))
	}

	httpPort := ":8480"
	if p := os.Getenv("HTTP_PORT"); p != "" {
		httpPort = p
	}
	app := newApp()

	// Load TLS assets
	cer, err := tls.LoadX509KeyPair("certs/dev.opsani.com+3.pem", "certs/dev.opsani.com+3-key.pem")
	if err != nil {
		log.Fatal(err)
	}

	config := &tls.Config{Certificates: []tls.Certificate{cer}}

	// Create TLS port listener
	httpsPort := ":8843"
	if p := os.Getenv("HTTPS_PORT"); p != "" {
		httpsPort = p
	}
	ln, err := tls.Listen("tcp", httpsPort, config)
	if err != nil {
		panic(err)
	}

	// Listen with TLS on HTTPS_PORT (:8843)
	go func() {
		log.Fatal(app.Listener(ln))
	}()

	// Listen on HTTP_PORT (:8480)
	log.Fatal(app.Listen(httpPort))
}

// NewRelicMiddleware instruments the request with New Relic
func NewRelicMiddleware(app *newrelic.Application) fiber.Handler {
	return func(c *fiber.Ctx) error {
		// start an HTTP transaction with New Relic
		txn := app.StartTransaction(c.Path())
		defer txn.End()

		// let Fiber process the request
		c.Next()

		// translate the FastHTTP request & response for New Relic
		hdr := make(http.Header)
		c.Context().Request.Header.VisitAll(func(k, v []byte) {
			sk := string(k)
			sv := string(v)
			hdr.Set(sk, sv)
		})

		txn.SetWebRequest(newrelic.WebRequest{
			Header:    http.Header{},
			URL:       &url.URL{Path: c.Path()},
			Method:    c.Method(),
			Transport: newrelic.TransportHTTP,
		})

		// Get a New Relic wrapper for the response writer
		rw := txn.SetWebResponse(nil)
		rw.WriteHeader(c.Context().Response.StatusCode())
		_, err := rw.Write(c.Context().Response.Body())
		return err
	}
}
